### ISSUE TITLE
[RFC]: Implement return indexer topk for sparse attention.

### ISSUE THREAD

**Issue by Ronald1995** [vllm-project/vllm#47280]
### Motivation.

DeepSeek Sparse Attention (DSA) models (DeepSeek-V3.2, V4, GLM-5.1-MoE,
MiniMax-M3) compute a top-k token selection at each indexer layer: for every
query token, the indexer picks the k most-relevant KV slots that the
subsequent sparse-attention layer will attend to. These indices are a
first-class output of the model's internal computation. The primary
motivation for surfacing them to the client is **training-inference
consistency**:

1. **Top-k selection is part of the model's forward semantics, not just an
   inference optimization.** During training, the indexer's top-k decides
   which KV slots receive gradients via the sparse-attention path. If
   training uses one set of selected slots (e.g., computed in a custom
   training kernel or on full attention masks) while inference uses a
   different set (computed by vLLM's `sparse_attn_indexer` under
   KV-cache/prefix-cache/block-layout constraints), the deployed model's
   behavior silently diverges from the trained one. Returning the
   inference-time topk lets trainers diff `topk_train` vs `topk_vllm` and
   quantify / close the gap.

2. **KV-cache layout differences leak into top-k selection.** Training
   typically runs on contiguous `[seq_len, ...]` layouts; vLLM runs on
   paged/block-table layouts with prefix caching, chunked prefill, and
   async preemption. Even when the indexer math is identical, the set of
   candidate slots visible to the indexer can differ (e.g., cache-hit
   tokens reusing previously computed topk vs recomputing). Surfacing
   vLLM's actual topk makes these layout-induced divergences observable
   and auditable.

Today, vLLM computes these indices internally and discards them after each
forward pass. There is no API to retrieve them, so the only way to audit
training-inference topk consistency is to re-implement vLLM's indexer
externally — fragile and guaranteed to drift. SGLang has shipped
`enable_return_indexer_topk` for the same reason; this RFC brings parity
to vLLM.

The proposed design is the **minimal extension** of the proven
`enable_return_routed_experts` infrastructure:
the same data-flow shape, the same async-scheduling safety properties, and
the same API surface — only the captured tensor differs (topk indices vs
expert routing).

### Proposed Change.

```mermaid
flowchart TD
    subgraph Worker["Worker (GPU)"]
        A["SparseAttnIndexer.forward_cuda"]
        B["capture_fn(topk_indices)"]
        C["IndexerTopkCapturer.device_buffer<br/>GPU, [N, L_idx, k]"]
        D["_bookkeeping_sync /<br/>AsyncGPUModelRunnerOutput<br/>non-blocking D2H"]
        E["pinned CPU<br/>IndexerTopkLists (numpy)"]
        A --> B --> C --> D --> E
    end

    E -->|"ModelRunnerOutput.indexer_topk"| F

    subgraph Scheduler["Scheduler (CPU)"]
        F["IndexerTopkManager.store_batch<br/>topk_data, slot_mapping"]
        G["indexer_topk_by_slot[slot_mapping] = topk_data<br/>slot = block_id * block_size + offset<br/>keyed by KV-cache layout"]
        H["On request finish:<br/>IndexerTopkManager.get<br/>block_ids, num_tokens, prompt_start"]
        I["per-request ndarray<br/>[num_tokens, L_idx, k]"]
        F --> G
        G -.->|"read on finish"| H --> I
    end

    I -->|"EngineCoreOutput.indexer_topk"| J

    subgraph API["OutputProcessor → API"]
        J["output_processor:<br/>concatenate chunks<br/>(prompt + decode)"]
        K["CompletionOutput.indexer_topk<br/>(ndarray)"]
        L["OpenAI serving:<br/>base64(np.save(.npy)) string"]
        M["response.choices[0].indexer_topk"]
        J --> K --> L --> M
    end

    classDef gpu stroke:#6b8cff,stroke-width:2px
    classDef cpu stroke:#36b37e,stroke-width:2px
    classDef api stroke:#ff8b3d,stroke-width:2px
    class A,B,C,D gpu
    class E,F,G,H,I cpu
    class J,K,L,M api
```

### Any Other Things.

these is my draft implementation: https://github.com/vllm-project/vllm/pull/47279

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
